lib/update.go

```diff
@@ -0,0 +1,399 @@
+	if err != nil {
+		return err
+	}
+
+	*res = jobs
+	return nil
+}
+
+// LogFile reads log file data for a given logName
+func (m *UpdateMethods) LogFile(logName *string, data *[]byte) error {
+	f, err := m.inst.cron.LoggedJobFile(context.Background(), *logName)
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+	res, err := ioutil.ReadAll(f)
+	*data = res
+
+	return err
+}
+
+// ServiceStatus describes the current state of a service
+type ServiceStatus struct {
+	Name       string
+	Running    bool
+	Daemonized bool // if true this service is scheduled
+	Started    *time.Time
+	Address    string
+	Metrics    map[string]interface{}
+}
+
+// ServiceStatus reports status of the cron daemon
+func (m *UpdateMethods) ServiceStatus(in *bool, out *ServiceStatus) error {
+	return fmt.Errorf("not finished")
+}
+
+// UpdateServiceStart starts the update service
+func UpdateServiceStart(ctx context.Context, repoPath string, updateCfg *config.Update, opts []Option) error {
+	if updateCfg == nil {
+		updateCfg = config.DefaultUpdate()
+	}
+
+	cli := cron.HTTPClient{Addr: updateCfg.Address}
+	if err := cli.Ping(); err == nil {
+		return fmt.Errorf("service already running")
+	}
+
+	var jobStore, logStore cron.JobStore
+	switch updateCfg.Type {
+	case "fs":
+		jobStore = cron.NewFlatbufferJobStore(repoPath + "/cron_jobs.qfb")
+		logStore = cron.NewFlatbufferJobStore(repoPath + "/cron_logs.qfb")
+	case "mem":
+		jobStore = &cron.MemJobStore{}
+		logStore = &cron.MemJobStore{}
+	default:
+		return fmt.Errorf("unknown cron type: %s", updateCfg.Type)
+	}
+
+	svc := cron.NewCron(jobStore, logStore, updateFactory)
+	log.Debug("starting update service")
+	go func() {
+		if err := svc.ServeHTTP(updateCfg.Address); err != nil {
+			log.Errorf("starting cron http server: %s", err)
+		}
+	}()
+
+	return svc.Start(ctx)
+}
+
+// ServiceStart ensures the scheduler is running
+func (m *UpdateMethods) ServiceStart(in, out *bool) error {
+	local, ok := m.inst.cron.(*cron.Cron)
+	if !ok {
+		return fmt.Errorf("service already running")
+	}
+
+	ctx := context.Background()
+	if err := local.Start(ctx); err != nil {
+		return err
+	}
+	return nil
+}
+
+// ServiceStop halts the scheduler
+func (m *UpdateMethods) ServiceStop(in, out *bool) error {
+	// TODO (b5):
+	return fmt.Errorf("not finished")
+}
+
+// ServiceRestart uses shell commands to restart the scheduler service
+func (m *UpdateMethods) ServiceRestart(in, out *bool) error {
+	// TODO (b5):
+	return fmt.Errorf("not finished")
+}
+
+// Run advances a dataset to the latest known version from either a peer or by
+// re-running a transform in the peer's namespace
+func (m *UpdateMethods) Run(p *Job, res *repo.DatasetRef) (err error) {
+	if m.inst.rpc != nil {
+		return m.inst.rpc.Call("UpdateMethods.Run", p, res)
+	}
+
+	switch p.Type {
+	case cron.JTDataset:
+		params := &SaveParams{
+			Ref: p.Name,
+		}
+		if o, ok := p.Options.(*cron.DatasetOptions); ok {
+			params = &SaveParams{
+				Ref:                 p.Name,
```

Comment:
Wondering if this field should be renamed to Ref on DatasetOptions?

Similar to Remotes, we're running into a problem here where cron wants to be using `lib`'s params struct, but can't because it lives below `lib`'s level. In Remotes, we have to send an http request to another node, but don't have the params struct because it lives below `api`, so we have to copy it in two places.

Maybe the answer is the common pattern of dumb data: move the communication structs down as far as possible, so that anything can depend on them. That makes it a little tricky for `lib`'s outfacing interface, since we want people to be consuming `lib` directly, but maybe that can be solved by using type aliasing.

Reply:
> Wondering if this field should be renamed to Ref on DatasetOptions?

"Name" is intended to cover the ambiguity between scheduled shell script & dataset updates, the `Name` value is coming from the job, whereas all other values in this statement are coming from the `cron.DatasetOptions` pointer.

>  maybe that can be solved by using type aliasing.

Totally agreed. Starting to feel like an emerging pattern here. I'm wondering if those structs should be moved into base, then aliased in lib. Let's noodle on it.